File: server/src/services/aiClients.ts
import axios from "axios";
import { OpenAI } from "openai";
import { env } from "../config/env.js";

export type GenerateInput = {
  prompt: string;
  system: string;
  maxTokens?: number;
};

export type EmbeddingInput = {
  inputs: string[];
};

export interface LlmProvider {
  generate(input: GenerateInput): Promise<string>;
}

export interface EmbeddingProvider {
  embed(input: EmbeddingInput): Promise<number[][]>;
}

class OpenAiProvider implements LlmProvider, EmbeddingProvider {
  private client: OpenAI;

  constructor(apiKey: string) {
    this.client = new OpenAI({ apiKey });
  }

  async generate(input: GenerateInput): Promise<string> {
    const response = await this.client.chat.completions.create({
      model: env.aiModel,
      temperature: 0.2,
      max_tokens: input.maxTokens ?? 1024,
      messages: [
        { role: "system", content: input.system },
        { role: "user", content: input.prompt },
      ],
    });

    return (response.choices[0].message.content ?? "").trim();
  }

  async embed(input: EmbeddingInput): Promise<number[][]> {
    const response = await this.client.embeddings.create({
      model: env.embedModel,
      input: input.inputs,
    });
    return response.data.map((item) => item.embedding as number[]);
  }
}

class OllamaProvider implements LlmProvider, EmbeddingProvider {
  private baseUrl: string;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  async generate(input: GenerateInput): Promise<string> {
    const response = await axios.post(`${this.baseUrl}/api/generate`, {
      model: env.aiModel,
      prompt: `${input.system}\n\n${input.prompt}`,
      stream: false,
    });
    return (response.data?.response ?? "").trim();
  }

  async embed(input: EmbeddingInput): Promise<number[][]> {
    const response = await axios.post(`${this.baseUrl}/api/embeddings`, {
      model: env.embedModel,
      input: input.inputs,
    });
    const embeddings = response.data?.embeddings ?? [];
    return embeddings as number[][];
  }
}

export type AiClients = {
  llm: LlmProvider;
  embed: EmbeddingProvider;
};

export const createAiClients = (): AiClients => {
  const provider = env.aiProvider.toLowerCase();
  const embedProvider = env.embedProvider.toLowerCase();

  if (provider === "openai" && !env.openAiKey) {
    throw new Error("OPENAI_API_KEY is required when AI_PROVIDER=openai");
  }
  if (embedProvider === "openai" && !env.openAiKey) {
    throw new Error("OPENAI_API_KEY is required when EMBED_PROVIDER=openai");
  }

  const llm = provider === "ollama"
    ? new OllamaProvider(process.env.AI_BASE_URL ?? "http://localhost:11434")
    : new OpenAiProvider(env.openAiKey);

  const embed = embedProvider === "ollama"
    ? new OllamaProvider(process.env.EMBED_BASE_URL ?? "http://localhost:11434")
    : new OpenAiProvider(env.openAiKey);

  return { llm, embed };
};

File: apps/api/src/ai_clients.py
import json
from abc import ABC, abstractmethod

import httpx
from openai import AsyncOpenAI

from src.config import get_settings
from src.logging_setup import get_logger

settings = get_settings()
logger = get_logger("ai-clients")


class LlmProvider(ABC):
    @abstractmethod
    async def generate(self, prompt: str, system: str, max_tokens: int = 1024) -> str: ...

    @abstractmethod
    async def generate_json(self, prompt: str, system: str, max_tokens: int = 1024, schema: dict = None) -> dict: ...

    @abstractmethod
    async def generate_json_deterministic(
        self, prompt: str, system: str, max_tokens: int = 1024, schema: dict = None
    ) -> dict: ...


class EmbeddingProvider(ABC):
    @abstractmethod
    async def embed(self, inputs: list[str]) -> list[list[float]]: ...


class OpenAiProvider(LlmProvider, EmbeddingProvider):
    def __init__(self, api_key: str, base_url: str = None):
        kwargs = {"api_key": api_key}
        if base_url:
            kwargs["base_url"] = base_url
        self.client = AsyncOpenAI(**kwargs)

    async def generate(self, prompt: str, system: str, max_tokens: int = 1024) -> str:
        resp = await self.client.chat.completions.create(
            model=settings.ai_model,
            temperature=0.2,
            max_tokens=max_tokens,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": prompt},
            ],
        )
        return (resp.choices[0].message.content or "").strip()

    async def generate_json(self, prompt: str, system: str, max_tokens: int = 1024, schema: dict = None) -> dict:
        response_format = {"type": "json_object"}
        if schema:
            response_format = {
                "type": "json_schema",
                "json_schema": schema,
            }
        resp = await self.client.chat.completions.create(
            model=settings.ai_model,
            temperature=0.2,
            max_tokens=max_tokens,
            response_format=response_format,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": prompt},
            ],
        )
        content = (resp.choices[0].message.content or "").strip()
        try:
            return json.loads(content)
        except json.JSONDecodeError:
            raise ValueError("OpenAI returned invalid JSON")

    async def generate_json_deterministic(
        self, prompt: str, system: str, max_tokens: int = 1024, schema: dict = None
    ) -> dict:
        response_format = {"type": "json_object"}
        if schema:
            response_format = {
                "type": "json_schema",
                "json_schema": schema,
            }
        resp = await self.client.chat.completions.create(
            model=settings.ai_model,
            temperature=0,
            top_p=0.1,
            max_tokens=max_tokens,
            response_format=response_format,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": prompt},
            ],
        )
        content = (resp.choices[0].message.content or "").strip()
        try:
            return json.loads(content)
        except json.JSONDecodeError:
            raise ValueError("OpenAI returned invalid JSON")

    async def embed(self, inputs: list[str]) -> list[list[float]]:
        resp = await self.client.embeddings.create(
            model=settings.embed_model,
            input=inputs,
        )
        return [item.embedding for item in resp.data]


class OllamaProvider(LlmProvider, EmbeddingProvider):
    def __init__(self, base_url: str, is_embed: bool = False):
        self.base_url = base_url.rstrip("/")
        self.is_embed = is_embed

    async def generate(self, prompt: str, system: str, max_tokens: int = 1024) -> str:
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{self.base_url}/api/generate",
                json={
                    "model": settings.ai_model if not self.is_embed else settings.embed_model,
                    "prompt": f"{system}\n\n{prompt}",
                    "stream": False,
                },
            )
            resp.raise_for_status()
            return (resp.json().get("response") or "").strip()

    async def generate_json(self, prompt: str, system: str, max_tokens: int = 1024, schema: dict = None) -> dict:
        format_param = schema.get("schema") if schema else "json"
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{self.base_url}/api/generate",
                json={
                    "model": settings.ai_model if not self.is_embed else settings.embed_model,
                    "prompt": f"{system}\n\n{prompt}",
                    "stream": False,
                    "format": format_param,
                },
            )
            resp.raise_for_status()
            content = (resp.json().get("response") or "").strip()
            try:
                return json.loads(content)
            except json.JSONDecodeError:
                raise ValueError("Ollama returned invalid JSON")

    async def generate_json_deterministic(
        self, prompt: str, system: str, max_tokens: int = 1024, schema: dict = None
    ) -> dict:
        format_param = schema.get("schema") if schema else "json"
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{self.base_url}/api/generate",
                json={
                    "model": settings.ai_model if not self.is_embed else settings.embed_model,
                    "prompt": f"{system}\n\n{prompt}",
                    "stream": False,
                    "format": format_param,
                    "options": {
                        "temperature": 0,
                        "top_p": 0.1,
                    },
                },
            )
            resp.raise_for_status()
            content = (resp.json().get("response") or "").strip()
            try:
                return json.loads(content)
            except json.JSONDecodeError:
                raise ValueError("Ollama returned invalid JSON")

    async def embed(self, inputs: list[str]) -> list[list[float]]:
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{self.base_url}/api/embed",
                json={
                    "model": settings.embed_model,
                    "input": inputs,
                },
            )
            resp.raise_for_status()
            return resp.json().get("embeddings") or []


class AiClients:
    def __init__(self, llm: LlmProvider, embed: EmbeddingProvider):
        self.llm = llm
        self.embed = embed


def create_ai_clients() -> AiClients:
    ai_provider = settings.ai_provider.lower()
    embed_provider = settings.embed_provider.lower()

    if ai_provider == "openai" and not settings.openai_api_key:
        raise ValueError("OPENAI_API_KEY is required when AI_PROVIDER=openai")
    if embed_provider == "openai" and not settings.openai_api_key:
        raise ValueError("OPENAI_API_KEY is required when EMBED_PROVIDER=openai")

    if ai_provider == "ollama":
        llm = OllamaProvider(settings.ai_base_url)
    else:
        llm = OpenAiProvider(settings.openai_api_key, base_url=settings.openai_api_base)

    if embed_provider == "ollama":
        embed = OllamaProvider(settings.embed_base_url, is_embed=True)
    else:
        embed_base = settings.embed_api_base or settings.openai_api_base
        embed = OpenAiProvider(settings.openai_api_key, base_url=embed_base)

    return AiClients(llm=llm, embed=embed)
